HPCC-17402 WsDfu/ListHistory & WsDfu/EraseHistory to return a validU/I response

AttilaVamos opened the pull request:
Add history record strucute and new array of history records data member for ListHistoryResponse and EraseHistoryResponse
    
Deprecate old binary xmlmap data member
    
Implement history tree to history array conversion in DFU Service
    
Modify WsDfuEx::onEraseHistory() and WsDfuEx::onListHistory() methods to provide history array instead of xmlmap
    
Modify DFUPlus listhistory() and erasehistory() methods to process array of history records instead of xmlmap to list/back up the file history.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 5ef8f6b4992aae2381089727341f25c23b63ebe4
Build: failed
Number of Error(s): 7
/mnt/disk1/home/vamosax/smoketest/PR-9865/HPCC-Platform/dali/dfuplus/dfuplus.cpp:1787:33: error: \xe2\x80\x98arrHistory\xe2\x80\x99 was not declared in this scope
make[2]: *** [dali/dfuplus/CMakeFiles/dfuplus.dir/dfuplus.cpp.o] Error 1
/mnt/disk1/home/vamosax/smoketest/PR-9865/HPCC-Platform/dali/dfuplus/dfuplus.cpp:1787:33: error: \xe2\x80\x98arrHistory\xe2\x80\x99 was not declared in this scope
make[2]: *** [plugins/fileservices/CMakeFiles/fileservices.dir/__/__/dali/dfuplus/dfuplus.cpp.o] Error 1
make[1]: *** [dali/dfuplus/CMakeFiles/dfuplus.dir/all] Error 2
make[1]: *** [plugins/fileservices/CMakeFiles/fileservices.dir/all] Error 2
make: *** [all] Error 2
Get ECLWatch build result

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: f0ef4c069448a53cfd0b5766a40d95b50b8896d3
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.572 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |722 |722 |0 |
|test (thor) |617 |617 |0 |
|test (roxie) |746 |746 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@afishbeck please review.

richardkchapman commented:
Checklist missing

AttilaVamos commented:
Sorry, updated.

rpastrana reviewed (changes requested):
@AttilaVamos a few comments on the ECM code

AttilaVamos commented on esp/scm/ws_dfu.ecm:
```diff
@@ -707,15 +707,26 @@ ESPresponse [exceptions_inline, nil_remove, http_encode(0)] DFUGetFileMetaDataRe
-
 ESPrequest ListHistoryRequest
 {
     string name;
```
Changed

AttilaVamos commented on esp/scm/ws_dfu.ecm:
```diff
@@ -707,15 +707,26 @@ ESPresponse [exceptions_inline, nil_remove, http_encode(0)] DFUGetFileMetaDataRe
 };
 
+ESPStruct FileHistory
+{
```
Changed

AttilaVamos commented on esp/scm/ws_dfu.ecm:
```diff
@@ -707,15 +707,26 @@ ESPresponse [exceptions_inline, nil_remove, http_encode(0)] DFUGetFileMetaDataRe
 {
-    binary xmlmap;
+    [depr_ver("1.36")] binary xmlmap;
+    ESParray<ESPStruct FileHistory> History;
```
Changed

AttilaVamos commented on esp/scm/ws_dfu.ecm:
```diff
@@ -725,16 +736,17 @@ ESPrequest EraseHistoryRequest
 {
-    binary xmlmap;
+    [depr_ver("1.36")] binary xmlmap;
+    ESParray<ESPStruct FileHistory> History;
```
Changed

AttilaVamos commented:
@rpastrana Updated, please re-review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 3e31004889de30cc56d6b7f940594cc44af429e5
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.917 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |722 |722 |0 |
|test (thor) |617 |617 |0 |
|test (roxie) |746 |746 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1518,16 +1518,16 @@ int CDfuPlusHelper::listhistory()
-                    out.appendf("\t\ttimestamp=\"%s\"/>", historyRecord->getOriginTimestamp()).newline();
-                    out.appendf("\t\tworkunit=\"%s\"/>", historyRecord->getOriginWorkunit()).newline();
+                    out.appendf("\t<Origin ip=\"%s\"", historyRecord->getIP()).newline();
+                    out.appendf("\t\tname=\"%s\"", historyRecord->getName()).newline();
```
since you're already doing an appendf, it might be more efficient to add the new line in the formatted string as opposed to calling the newline method.
out.appendf("\t\tname=\"%s\"\n", historyRecord->getName());

AttilaVamos replied:
Changed

rpastrana commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1613,17 +1613,17 @@ int CDfuPlusHelper::listhistory()
-                    jsonDump.appendf("%s\"path\" : \"%s\",", level3, historyRecord->getOriginPath()).newline();
-                    jsonDump.appendf("%s\"timestamp\" : \"%s\",", level3, historyRecord->getOriginTimestamp()).newline();
-                    jsonDump.appendf("%s\"workunit\" : \"%s\"", level3, historyRecord->getOriginWorkunit()).newline();
+                    jsonDump.appendf("%s\"ip\" : \"%s\"," ,level3, historyRecord->getIP()).newline();
```
Same new line comment as above.

AttilaVamos replied:
Changed

rpastrana commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1678,16 +1678,16 @@ int CDfuPlusHelper::erasehistory()
-                out.appendf("\t\tpath=\"%s\"", historyRecord->getOriginPath()).newline();
-                out.appendf("\t\ttimestamp=\"%s\"/>", historyRecord->getOriginTimestamp()).newline();
-                out.appendf("\t\tworkunit=\"%s\"/>", historyRecord->getOriginWorkunit()).newline();
+                out.appendf("\t<Origin ip=\"%s\"", historyRecord->getIP()).newline();
```
Same new line comment as above.

AttilaVamos replied:
Changed

rpastrana commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -4988,20 +4988,20 @@ bool CWsDfuEx::onDFUBrowseData(IEspContext &context, IEspDFUBrowseDataRequest &r
-        historyRecord->setOriginPath(historyIter->query().queryProp("@path"));
-        historyRecord->setOriginTimestamp(historyIter->query().queryProp("@timestamp"));
-        historyRecord->setOriginWorkunit(historyIter->query().queryProp("@workunit"));
+        historyRecord->setIP(historyIter->query().queryProp("@ip"));
```
Assign historyIter->query() to a local var, and queryProp directly

AttilaVamos replied:
Changed

rpastrana reviewed (commented):
@AttilaVamos ECM changes look good, noticed a couple of other issues

rpastrana commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1641,20 +1667,46 @@ int CDfuPlusHelper::erasehistory()
+            IConstHistory* historyRecord = &arrHistory.item(i);
+            if(historyRecord != NULL)
+            {
+                out.appendf("\t<Origin ip=\"%s\"", historyRecord->getIP()).newline();
```
why do we use tabs for xml but 3 spaces for JSON, is that part of a standard? if not we might want to stay consistent.

rpastrana commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1641,20 +1667,46 @@ int CDfuPlusHelper::erasehistory()
+        }
+        out.append("</History>").newline();
+
+        info("\n%s\n",out.trim().str());
```
the trim seems superfluous since we control 'out'

rpastrana replied:
minor... but do we need to do the trim?

rpastrana reviewed (commented):
@AttilaVamos a couple of more comments

AttilaVamos commented:
@rpastrana Updated, please re-review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: be9d13fea3faae9aaf959c1d24e267079272ffef
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.55 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |722 |722 |0 |
|test (thor) |617 |617 |0 |
|test (roxie) |746 |746 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana reviewed (commented):
@AttilaVamos looks good. One minor question regarding a trim call on a string we construct.

@afishbeck can you take a quick look? thanks.

AttilaVamos commented:
@rpastrana the last few issues are addressed in my latest push. Please check.

rpastrana commented:
@AttilaVamos looks good. 
@afishbeck is going to take a quick look as well.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 103fc6f6a2271ebe316c6a97d04f17029dd09612
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.565 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |724 |724 |0 |
|test (thor) |619 |619 |0 |
|test (roxie) |747 |747 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1503,112 +1503,137 @@ int CDfuPlusHelper::listhistory()
+                if(historyRecord != nullptr)
+                {
+                    // Historical record item object
+                    jsonDump.appendf("%s\"ip\" : \"%s\",\n" ,level3, historyRecord->getIP());
```
There are functions to make this easier.
For example:
  appendJSONValue(jsonDump, "ip", historyRecord->getIP());

Hides all of the formatting so it's easier to read, you just need to add padding.

AttilaVamos replied:
Changed

afishbeck commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1641,20 +1666,46 @@ int CDfuPlusHelper::erasehistory()
-        history.setown(createPTree((MemoryBuffer &)respMsg));
+        StringBuffer out;
+        out.append("<History>\n");
+        for(unsigned i = 0; i < arrHistory.length(); i++)
```
Can this be common code with list history above?

AttilaVamos replied:
Done

afishbeck commented on esp/scm/ws_dfu.ecm:
```diff
@@ -707,34 +707,46 @@ ESPresponse [exceptions_inline, nil_remove, http_encode(0)] DFUGetFileMetaDataRe
 {
-    binary xmlmap;
+    [depr_ver("1.36")] binary xmlmap;
+    ESParray<ESPStruct History> FileHistory;
```
For consistency with other places shouldn't the item name be "Origin"?  And maybe the tag name "History"? but FileHistory is ok.

AttilaVamos replied:
It could be, but you should come to a common ground with @rpastrana what you want and record it into the related JIRA.

rpastrana replied:
@afishbeck the items in the repeating element are actually describing a file operation, so what we're reporting is a history of file operations taken upon the given file, what about:
ESParray<ESPStruct Operation> Operations; or
ESParray<ESPStruct Operation> History;

afishbeck replied:
@rpastrana @AttilaVamos It wasn't the tag name that bothered me, it was the item name.  History doesn't seem like the name of a single operation.  I also thought the web service should be consistent with the naming of the command line output.

afishbeck replied:
On the command line it is "ESParray<ESPStruct Origin> History" why the inconsistency?

afishbeck replied:
@rpastrana I probably can't read your comment as intended.  Anything in angle brackets disappears if you don't mark it up.

rpastrana replied:
oh ok, the inconsistency has to be addressed, I thought you were pointing out the element names weren't optimal.

If you're ok with `ESParray<ESPStruct History> FileHistory;` then let's keep it, otherwise one of these two might be more appropriate:
`ESParray<ESPStruct Operation> Operations;` or
`ESParray<ESPStruct Operation> History;`

rpastrana replied:
@AttilaVamos based on the offline conversation, let's define the File history array as follows:

`ESParray<ESPStruct History, Origin> History;`

notice we're now referencing the ESPStruct 'History' explicitely. Also, make sure the command line uses the same format.

afishbeck commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5008,17 +5027,19 @@ bool CWsDfuEx::onListHistory(IEspContext &context, IEspListHistoryRequest &req,
         else
             throw MakeStringException(ECLWATCH_FILE_NOT_EXIST,"CWsDfuEx::onListHistory: Could not find file '%s'.", req.getName());
-
-        resp.setXmlmap(xmlmap);
```
For backward compatibility, if the client is using the old interface version, don't you need to set xmlmap?  Not sure which version the original response was added, 1.35?  You still want people using that client to be able to work.

AttilaVamos replied:
Only the DFUPlus used this interface before.

richardkchapman replied:
You don't know whether externally written tools have been using this interface...

AttilaVamos replied:
That's can be true, but I introduced this interface in 6.2. 
BTW: Ok, I will put back that line.

AttilaVamos replied:
Done

afishbeck reviewed (changes requested):
@AttilaVamos some comments.

AttilaVamos commented:
@afishbeck updated, please re-review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d60c48d5bfbfb4d69c36831326305bef7a452fac
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.326 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |726 |726 |0 |
|test (thor) |619 |619 |0 |
|test (roxie) |747 |747 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
I will squash it after @afishbeck checks my latest update.

afishbeck commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5033,13 +5033,18 @@ bool CWsDfuEx::onListHistory(IEspContext &context, IEspListHistoryRequest &req,
             if (history)
+            {
                 storeHistoryTreeToArray(history, arrHistory);
+                history->serialize(xmlmap);
```
Sorry, I should have explained further.  Just setting the field is fine for anything that is not too expensive.  Versioning will remove it automatically if needed.  But for something large or expensive you can check the version and only set it if needed.

`if (context.getClientVersion() < 1.36)`
`   history->serialze(xmlmap)`;
`if (xmlmap.length())`
`   resp.setXmlmap(xmlmap);`

AttilaVamos replied:
Changed.

afishbeck commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1503,112 +1524,135 @@ int CDfuPlusHelper::listhistory()
+            for(unsigned i = 0; i < arrHistory.length(); i++)
+            {
+                if (index++ != 1)
+                    jsonDump.append(",");
```
This is fine, but I just want to mention if you use `delimitJSON(jsonDump);` you don't need the if.  Just looks slightly simpler.

AttilaVamos replied:
I think I will keep my solution.

afishbeck commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1503,112 +1524,135 @@ int CDfuPlusHelper::listhistory()
+                    // Historical record item object
+                    jsonDump.append(level3);
+                    appendJSONValue(jsonDump, "ip", historyRecord->getIP());
+                    jsonDump.appendf(",\n%s",level3);
```
Btw, another option is `delimitJSON(jsonDump, true).append(level3);`

AttilaVamos replied:
OK, what is your final suggestion?

AttilaVamos replied:
I think I will keep my solution for this.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: f3da10e0b849c7add01825384103a3bc053b3226
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.553 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |726 |726 |0 |
|test (thor) |619 |619 |0 |
|test (roxie) |747 |747 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@rpastrana  @afishbeck next iteration, please check.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 9d3edfb4d040730d43b810df68dd270e541bfb89
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.792 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |726 |726 |0 |
|test (thor) |619 |619 |0 |
|test (roxie) |747 |747 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented:
My concerns were addressed. back to @afishbeck

afishbeck commented on dali/dfuplus/dfuplus.cpp:
```diff
@@ -1456,6 +1456,27 @@ int CDfuPlusHelper::abort()
     return 0;
 }
 
+StringBuffer & historyToXml(IArrayOf<IConstHistory>& arrHistory, StringBuffer & out)
```
arrHistory could be const.

AttilaVamos replied:
Changed

afishbeck reviewed (approved):
Other than minor comment, looks fine.

AttilaVamos committed:
HPCC-17402 WsDfu/ListHistory & WsDfu/EraseHistory to return a valid U/I response

Add history record strucute and new array of history records data member for
ListHistoryResponse and EraseHistoryResponse

Deprecate old binary xmlmap data member

Implement history tree to history array conversion in DFU Service

Modify WsDfuEx::onEraseHistory() and WsDfuEx::onListHistory() methods to
provide history array instead of xmlmap

Modify DFUPlus listhistory() and erasehistory() methods to process array
of history records instead of xmlmap to list/back up the file history.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 6d1b1a3350563f0dec2bbe096daa1af8abb393c0
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.438 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |726 |726 |0 |
|test (thor) |619 |619 |0 |
|test (roxie) |747 |747 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@richardkchapman updated, squashed, please merge.
